### PR TITLE
Make JsonReader.promoteNameToValue() a public API

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

**Comment by JakeWharton** [comment]
What about `promoteValueToName`? Seems like we should do both or none.

**Comment by swankjesse** [comment]
We’ve got one use case for `promoteValueToName()`: it lets you reuse existing `JsonAdapters` for map keys without their knowledge. Moshi uses it internally for things like maps with enum keys.

I can’t think of a use case for `promoteValueToName`.

**Comment by swankjesse** [comment]
We’ve got one use case for promoteNameToValue(): it lets you reuse existing JsonAdapters for map keys without their knowledge. Moshi uses it internally for things like maps with enum keys.

I can’t think of a use case for promoteValueToName().

**Comment by JakeWharton** [comment]
Isn't it the symmetric same use case? Using an adapter to write a value as a name.

**Comment by swankjesse** [comment]
Yeah but because we need the value case much much more we can solve both cases with only this one.

**Comment by JakeWharton** [comment]
I don't follow. How do I write an adapter for a `Map<T, String>` given only an opaque adapter for `T`?

**Comment by swankjesse** [comment]
Ahhhhhhhh, yep I totally missed this. I was worrying all about reading these things that I forgot all about writing ’em. That is now fixed, please take another look!

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java [hunk 1]
```diff
@@ -514,8 +514,19 @@ public final void setFailOnUnknown(boolean failOnUnknown) {
   /**
    * Changes the reader to treat the next name as a string value. This is useful for map adapters so
    * that arbitrary type adapters can use {@link #nextString} to read a name value.
+   *
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java:517]: suuuuuper nit: this is Java and you need to take a `<p>` 😢

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.java [hunk 2]
```diff
@@ -443,8 +443,20 @@ public final JsonWriter jsonValue(@Nullable Object value) throws IOException {
   /**
    * Changes the writer to treat the next value as a string name. This is useful for map adapters so
    * that arbitrary type adapters can use {@link #value} to write a name value.
+   *
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonWriter.java:446]: Also `<p>`

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java (hunk 1 above)
Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonReader.java:517]: OMG how quickly I forget!

File: moshi/src/main/java/com/squareup/moshi/JsonWriter.java (hunk 2 above)
Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonWriter.java:446]: ℙ

**Pull request merged**